{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4440cd32-3c8a-4163-aaf4-2c06fd600538",
   "metadata": {},
   "source": [
    "# Simple example\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f103a80-961f-4ed2-b5d4-61bba207a0c6",
   "metadata": {},
   "source": [
    "## Data download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7c19c738-f876-4078-9202-f70f28041e29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gdown in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (4.7.1)\n",
      "Requirement already satisfied: requests[socks] in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from gdown) (2.26.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from gdown) (4.10.0)\n",
      "Requirement already satisfied: six in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from gdown) (1.16.0)\n",
      "Requirement already satisfied: tqdm in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from gdown) (4.62.3)\n",
      "Requirement already satisfied: filelock in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from gdown) (3.3.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from beautifulsoup4->gdown) (2.2.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from requests[socks]->gdown) (1.26.7)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from requests[socks]->gdown) (2.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from requests[socks]->gdown) (2021.10.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from requests[socks]->gdown) (2.0.7)\n",
      "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from requests[socks]->gdown) (1.7.1)\n",
      "Requirement already satisfied: lightgbm in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (4.0.0)\n",
      "Requirement already satisfied: numpy in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from lightgbm) (1.21.2)\n",
      "Requirement already satisfied: scipy in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from lightgbm) (1.7.2)\n",
      "Requirement already satisfied: xgboost in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (1.7.5)\n",
      "Requirement already satisfied: numpy in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from xgboost) (1.21.2)\n",
      "Requirement already satisfied: scipy in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from xgboost) (1.7.2)\n",
      "Requirement already satisfied: catboost==1.1.1 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (1.1.1)\n",
      "Requirement already satisfied: pandas>=0.24.0 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (1.3.4)\n",
      "Requirement already satisfied: graphviz in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (0.20.1)\n",
      "Requirement already satisfied: scipy in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (1.7.2)\n",
      "Requirement already satisfied: numpy>=1.16.0 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (1.21.2)\n",
      "Requirement already satisfied: six in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (1.16.0)\n",
      "Requirement already satisfied: matplotlib in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (3.4.3)\n",
      "Requirement already satisfied: plotly in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from catboost==1.1.1) (5.15.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from pandas>=0.24.0->catboost==1.1.1) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from pandas>=0.24.0->catboost==1.1.1) (2021.3)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from matplotlib->catboost==1.1.1) (3.0.4)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from matplotlib->catboost==1.1.1) (0.10.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from matplotlib->catboost==1.1.1) (8.4.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from matplotlib->catboost==1.1.1) (1.3.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from plotly->catboost==1.1.1) (8.2.2)\n",
      "Requirement already satisfied: packaging in /Users/4pz/opt/anaconda3/lib/python3.8/site-packages (from plotly->catboost==1.1.1) (21.0)\n"
     ]
    }
   ],
   "source": [
    "# Download dependencies\n",
    "!pip install gdown\n",
    "\n",
    "!pip install lightgbm\n",
    "!pip install xgboost\n",
    "!pip install catboost==1.1.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b931ef01-4633-485e-a338-0995bc1c14fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File data/allgenomes.RAST.txt.ps.tsv already exists\n",
      "File data/allgenomes.interpro.txt.ps.tsv already exists\n"
     ]
    }
   ],
   "source": [
    "# Download both interpro and rast annotations for genomes for this to work. Stored in data/ Both files are in .gitignore\n",
    "import os\n",
    "\n",
    "\n",
    "\n",
    "# Check and download allgenomes.RAST.txt.ps.tsv file\n",
    "if not os.path.exists(\"data/allgenomes.RAST.txt.ps.tsv\"):\n",
    "    print(\"Downloading File data/allgenomes.RAST.txt.ps.tsv\")\n",
    "    !gdown --no-check-certificate 15py4qVFmLiHF-a341e_YX1w3HgEDwbDz -O data/allgenomes.RAST.txt.ps.tsv\n",
    "else:\n",
    "    print(\"File data/allgenomes.RAST.txt.ps.tsv already exists\")\n",
    "\n",
    "# Check and download allgenomes.interpro.txt.ps.tsv file\n",
    "if not os.path.exists(\"data/allgenomes.interpro.txt.ps.tsv\"):\n",
    "    print(\"Downloading File data/allgenomes.interpro.txt.ps.tsv\")\n",
    "    !gdown --no-check-certificate 1PuDgDnWd5qumI0hkIKNYrTqICnh-_WlS -O data/allgenomes.interpro.txt.ps.tsv\n",
    "else:\n",
    "    print(\"File data/allgenomes.interpro.txt.ps.tsv already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9564b8f8-57f5-4181-a542-dc08e1baa5ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SSO_dictionary.json                   gram_stain_bacdive.tsv\n",
      "allgenomes.RAST.txt.ps.tsv            metabolic_phenotype_data_bacdive.tsv\n",
      "allgenomes.interpro.txt.ps.tsv\n"
     ]
    }
   ],
   "source": [
    "ls data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1acd9808-dc54-4c8f-ae7a-c4ee547b0cdd",
   "metadata": {},
   "source": [
    "## Load genomic feature data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9fab5fc-3b08-4fe8-880f-f966a4a27d71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#Load genomic features \n",
    "#Load rast features\n",
    "rast_df = pd.read_csv(\"data/allgenomes.RAST.txt.ps.tsv\", sep = \"\\t\", index_col=0)\n",
    "\n",
    "#Load interpro features\n",
    "interpro_df = pd.read_csv(\"data/allgenomes.interpro.txt.ps.tsv\", sep = \"\\t\", index_col=0)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53fb110b-b04a-444a-8e8e-77d016bd6b53",
   "metadata": {},
   "source": [
    "## A simple example with gram_stain data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2380e343-13f0-4f58-b4ff-1a5883ad3f7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phenotype: gram_stain\n",
      "Feature used: Rast\n",
      "Numer of genomes:3531\n",
      "Number of genomic features :6997\n",
      "Shape of y:(3531,)\n",
      "Accuracy = 0.9646393210749646\n",
      "precision = 0.9658944260240436\n",
      "Recall = 0.9646393210749646\n",
      "F1-score = 0.9648157949771774\n",
      "Balanced accuracy = 0.9679094313925775\n",
      "Confusion matrix:\n",
      "[[420  20]\n",
      " [  5 262]]\n"
     ]
    }
   ],
   "source": [
    "# Run a simple example\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, balanced_accuracy_score, confusion_matrix\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "#Load phenotype data with gram stain data (only gram stain phenotype here) and create dataset used\n",
    "\n",
    "phenotype_gram_stain_df = pd.read_csv(\"data/gram_stain_bacdive.tsv\", sep = \"\\t\", index_col=0)\n",
    "phenotype = \"gram_stain\"\n",
    "phenotype_df = phenotype_gram_stain_df[phenotype].dropna()\n",
    "\n",
    "#Merged genotype and phenotype data and extract X and y\n",
    "\n",
    "merged_df = rast_df.merge(phenotype_df, left_index=True, right_index=True, how='inner')\n",
    "X = merged_df.drop(columns=[phenotype])\n",
    "y = merged_df[phenotype]\n",
    "\n",
    "\n",
    "# print some info\n",
    "\n",
    "print (\"Phenotype: \" + phenotype)\n",
    "print (\"Feature used: Rast\")\n",
    "\n",
    "num_genome, num_features = X.shape\n",
    "print (\"Numer of genomes:\" + str(num_genome) + \"\\n\" + \"Number of genomic features :\" + str(num_features)) \n",
    "print (\"Shape of y:\" + str(y.shape))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Run random forest classifier\n",
    "\n",
    "# Split the data into training and testing sets\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=112)\n",
    "rfc = RandomForestClassifier(n_estimators=1000)\n",
    "rfc.fit(X_train, y_train)\n",
    "y_pred = rfc.predict(X_test)\n",
    "\n",
    "\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "precision = precision_score(y_test, y_pred, average='weighted')\n",
    "recall = recall_score(y_test, y_pred, average='weighted')\n",
    "f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "confusion_matrix_info = confusion_matrix(y_test, y_pred)\n",
    "\n",
    "\n",
    "print (\"Accuracy = \" + str(accuracy))\n",
    "print (\"precision = \" + str(precision))\n",
    "print (\"Recall = \" + str(recall))\n",
    "print (\"F1-score = \" + str(f1))\n",
    "print (\"Balanced accuracy = \" + str(balanced_accuracy))\n",
    "       \n",
    "    \n",
    "print (\"Confusion matrix:\")\n",
    "print (confusion_matrix_info)\n",
    "       \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a3a4714-71e7-4893-8512-4baa5eeed277",
   "metadata": {},
   "source": [
    "## More complicated example: Run multiple classifiers together, get top features and show html report\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32142eab-c46d-4cd6-be41-d60c6281a051",
   "metadata": {},
   "source": [
    "### Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "21f9dc1a-0207-4dcc-8750-a5fcb9569b40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import uuid\n",
    "def generate_html_table(df: pd.DataFrame):\n",
    "        \"\"\"Display a pandas.DataFrame as jQuery DataTables\"\"\"\n",
    "\n",
    "        # Generate random container name\n",
    "        id_container = uuid.uuid1()\n",
    "        output = \"\"\"\n",
    "    <div id=\"datatable-container-{id_container}\">\n",
    "      <script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.7.0/jquery.min.js\"></script>\n",
    "      <script type=\"text/javascript\" src=\"https://cdn.datatables.net/1.13.5/js/jquery.dataTables.min.js\"></script>\n",
    "      <link rel=\"stylesheet\" type=\"text/css\" href=\"https://cdn.datatables.net/1.13.5/css/jquery.dataTables.min.css\"/>\n",
    "      <script type=\"text/javascript\">\n",
    "        $(document).ready( function () {{\n",
    "            $('#BGCtable').DataTable();\n",
    "        }});\n",
    "      </script>\n",
    "      <!-- Insert table below -->\n",
    "      {table}\n",
    "    </div>\n",
    "        \"\"\".format(\n",
    "            id_container=id_container,\n",
    "            table=df.to_html(\n",
    "                index=False,\n",
    "                table_id=\"classification\",\n",
    "                classes=\"display\"\n",
    "            ),\n",
    "        )\n",
    "        return output\n",
    "    \n",
    "\n",
    "def get_classifier_report(X, y):\n",
    "\n",
    "    import numpy as np\n",
    "    import pandas as pd\n",
    "    from sklearn.model_selection import train_test_split\n",
    "    from sklearn.linear_model import LogisticRegression\n",
    "    from sklearn.tree import DecisionTreeClassifier\n",
    "    from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier\n",
    "    import xgboost as xgb\n",
    "    import lightgbm as lgb\n",
    "    import catboost as cb\n",
    "    from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, balanced_accuracy_score\n",
    "    from sklearn.feature_selection import SelectKBest, chi2\n",
    "    import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "    # Split the data into training and testing sets\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "    # Dictionary to store the classifiers and their parameters\n",
    "    classifiers = {\n",
    "\n",
    "        'Decision Tree': {\n",
    "            'model': DecisionTreeClassifier(),\n",
    "            'params': {}\n",
    "        },\n",
    "        'Random Forest': {\n",
    "            'model': RandomForestClassifier(),\n",
    "            'params': {'n_estimators': 1000, 'max_depth': None, 'random_state': 100}\n",
    "        },\n",
    "        'XGBoost': {\n",
    "            'model': xgb.XGBClassifier(),\n",
    "            'params': {'n_estimators': 1000, 'max_depth': 3, 'learning_rate': 0.1, 'random_state': 100}\n",
    "        },\n",
    "\n",
    "        'CatBoost': {\n",
    "            'model': cb.CatBoostClassifier(),\n",
    "            'params': {'iterations': 1000, 'depth': 6, 'learning_rate': 0.1, 'random_state': 42, 'verbose': False}\n",
    "        },\n",
    "        'AdaBoost': {\n",
    "            'model': AdaBoostClassifier(),\n",
    "            'params': {'n_estimators': 1000, 'random_state': 100}\n",
    "        }\n",
    "    }\n",
    "\n",
    "    html_table_rows = []\n",
    "\n",
    "\n",
    "\n",
    "    # Train and evaluate each classifier\n",
    "    for clf_name, clf_data in classifiers.items():\n",
    "        print (\"Running \" + clf_name)\n",
    "        model = clf_data['model']\n",
    "        params = clf_data['params']\n",
    "\n",
    "        # Train the classifier\n",
    "        model.set_params(**params)\n",
    "        model.fit(X_train, y_train)\n",
    "\n",
    "        # Make predictions on the test set\n",
    "        y_pred = model.predict(X_test)\n",
    "\n",
    "\n",
    "\n",
    "        accuracy = accuracy_score(y_test, y_pred)\n",
    "        precision = precision_score(y_test, y_pred, average='weighted')\n",
    "        recall = recall_score(y_test, y_pred, average='weighted')\n",
    "        f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "        balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "\n",
    "        # Extract top features (importances or coefficients)\n",
    "        try:\n",
    "            feature_importances = dict(zip(X.columns, model.feature_importances_))\n",
    "        except:\n",
    "            feature_importances = {}\n",
    "\n",
    "        # Sort and select the top 10 features\n",
    "        top_features = dict(sorted(feature_importances.items(), key=lambda x: x[1], reverse=True)[:20])\n",
    "\n",
    "        # Format the top features for prettier display\n",
    "        formatted_top_features = ', '.join([f\"{feat} ({importance:.2f})\" for feat, importance in top_features.items()])\n",
    "\n",
    "        # Append row to the HTML table\n",
    "        html_table_rows.append([clf_name, accuracy, precision, recall, f1, balanced_accuracy, formatted_top_features])\n",
    "\n",
    "    # Create a DataFrame to display the results\n",
    "    headers = [\"Model\", \"Accuracy\", \"Precision\", \"Recall\", \"F1-score\", \"Balanced Accuracy\", \"Top Features\"]\n",
    "    report_df = pd.DataFrame(html_table_rows, columns=headers)\n",
    "    return report_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d4dd1e-b84b-4686-803a-ab557333ff34",
   "metadata": {},
   "source": [
    "### Run multiple classifiers on gram stain data with RAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "be0c1f25-6700-4d6e-823d-e84d243e1fba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Decision Tree\n",
      "Running Random Forest\n",
      "Running XGBoost\n",
      "Running CatBoost\n",
      "Running AdaBoost\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div id=\"datatable-container-eb14e72c-308f-11ee-b959-8244d1051800\">\n",
       "      <script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.7.0/jquery.min.js\"></script>\n",
       "      <script type=\"text/javascript\" src=\"https://cdn.datatables.net/1.13.5/js/jquery.dataTables.min.js\"></script>\n",
       "      <link rel=\"stylesheet\" type=\"text/css\" href=\"https://cdn.datatables.net/1.13.5/css/jquery.dataTables.min.css\"/>\n",
       "      <script type=\"text/javascript\">\n",
       "        $(document).ready( function () {\n",
       "            $('#BGCtable').DataTable();\n",
       "        });\n",
       "      </script>\n",
       "      <!-- Insert table below -->\n",
       "      <table border=\"1\" class=\"dataframe display\" id=\"classification\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1-score</th>\n",
       "      <th>Balanced Accuracy</th>\n",
       "      <th>Top Features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.937765</td>\n",
       "      <td>0.937581</td>\n",
       "      <td>0.937765</td>\n",
       "      <td>0.937370</td>\n",
       "      <td>0.924944</td>\n",
       "      <td>SSO:000036573 (0.72), SSO:000024182 (0.10), SSO:000001995 (0.02), SSO:000043994 (0.01), SSO:000012980 (0.01), SSO:000020168 (0.01), SSO:000004595 (0.00), SSO:000043198 (0.00), SSO:000021874 (0.00), SSO:000042298 (0.00), SSO:000017481 (0.00), SSO:000019394 (0.00), SSO:000031056 (0.00), SSO:000020170 (0.00), SSO:000012708 (0.00), SSO:000022725 (0.00), SSO:000001091 (0.00), SSO:000019142 (0.00), SSO:000016805 (0.00), SSO:000016902 (0.00)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.973126</td>\n",
       "      <td>0.974158</td>\n",
       "      <td>0.973126</td>\n",
       "      <td>0.973287</td>\n",
       "      <td>0.976597</td>\n",
       "      <td>SSO:000007937 (0.03), SSO:000008104 (0.02), SSO:000006816 (0.02), SSO:000004460 (0.02), SSO:000036573 (0.02), SSO:000008604 (0.02), SSO:000043994 (0.01), SSO:000020865 (0.01), SSO:000000281 (0.01), SSO:000002241 (0.01), SSO:000009323 (0.01), SSO:000002230 (0.01), SSO:000004958 (0.01), SSO:000008598 (0.01), SSO:000008953 (0.01), SSO:000019343 (0.01), SSO:000021918 (0.01), SSO:000023921 (0.01), SSO:000004949 (0.01), SSO:000035277 (0.01)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XGBoost</td>\n",
       "      <td>0.964639</td>\n",
       "      <td>0.965278</td>\n",
       "      <td>0.964639</td>\n",
       "      <td>0.964789</td>\n",
       "      <td>0.965192</td>\n",
       "      <td>SSO:000036573 (0.40), SSO:000024182 (0.03), SSO:000006816 (0.03), SSO:000043994 (0.02), SSO:000002793 (0.02), SSO:000001995 (0.02), SSO:000008306 (0.02), SSO:000001301 (0.01), SSO:000042832 (0.01), SSO:000017976 (0.01), SSO:000008604 (0.01), SSO:000008104 (0.01), SSO:000008630 (0.01), SSO:000012980 (0.01), SSO:000008066 (0.01), SSO:000008135 (0.01), SSO:000018456 (0.01), SSO:000012951 (0.01), SSO:000019152 (0.01), SSO:000001853 (0.01)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CatBoost</td>\n",
       "      <td>0.966054</td>\n",
       "      <td>0.967032</td>\n",
       "      <td>0.966054</td>\n",
       "      <td>0.966243</td>\n",
       "      <td>0.968248</td>\n",
       "      <td>SSO:000036573 (28.02), SSO:000004460 (5.10), SSO:000043994 (4.39), SSO:000008104 (3.37), SSO:000006816 (3.00), SSO:000007937 (2.09), SSO:000024182 (1.98), SSO:000008604 (1.66), SSO:000001301 (1.48), SSO:000008105 (1.34), SSO:000002076 (1.21), SSO:000013395 (1.14), SSO:000004190 (1.05), SSO:000012160 (0.98), SSO:000002241 (0.89), SSO:000012811 (0.86), SSO:000000563 (0.81), SSO:000008306 (0.56), SSO:000006040 (0.53), SSO:000000281 (0.53)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AdaBoost</td>\n",
       "      <td>0.963225</td>\n",
       "      <td>0.963225</td>\n",
       "      <td>0.963225</td>\n",
       "      <td>0.963225</td>\n",
       "      <td>0.959162</td>\n",
       "      <td>SSO:000043092 (0.01), SSO:000008104 (0.01), SSO:000016911 (0.01), SSO:000018986 (0.01), SSO:000029838 (0.01), SSO:000012182 (0.01), SSO:000042756 (0.01), SSO:000001768 (0.01), SSO:000024895 (0.01), SSO:000001968 (0.01), SSO:000044120 (0.01), SSO:000008604 (0.01), SSO:000020269 (0.01), SSO:000012160 (0.01), SSO:000024618 (0.01), SSO:000020490 (0.01), SSO:000010833 (0.01), SSO:000025091 (0.01), SSO:000004100 (0.01), SSO:000005483 (0.01)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "    </div>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "\n",
    "#Load phenotype data with gram stain data (only gram stain phenotype here)\n",
    "phenotype_gram_stain_df = pd.read_csv(\"data/gram_stain_bacdive.tsv\", sep = \"\\t\", index_col=0)\n",
    "\n",
    "phenotype = \"gram_stain\"\n",
    "phenotype_df = phenotype_gram_stain_df[phenotype].dropna()\n",
    "merged_df = rast_df.merge(phenotype_df, left_index=True, right_index=True, how='inner')\n",
    "X = merged_df.drop(columns=[phenotype])\n",
    "y = merged_df[phenotype]\n",
    "\n",
    "# Get classifier report\n",
    "report_df = get_classifier_report(X,y)\n",
    "\n",
    "report_html_content = generate_html_table (report_df)\n",
    "\n",
    "HTML(report_html_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f4d28ed7-ce94-4933-9ca6-b58a7010a879",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SSO:000036573 (0.71)\tLPS-assembly protein LptD\n",
      "SSO:000024182 (0.10)\tRibosome maturation factor rimM\n",
      "SSO:000001301 (0.02)\tCDP-diacylglycerol--glycerol-3-phosphate 3-phosphatidyltransferase (EC 2.7.8.5)\n",
      "SSO:000012980 (0.01)\tDUF2225 domain-containing protein\n",
      "SSO:000031056 (0.01)\tmetalloenzyme domain protein\n",
      "SSO:000020884 (0.01)\tOuter membrane lipoprotein-sorting protein\n",
      "SSO:000043940 (0.01)\tdihydrolipoyllysine-residue acetyltransferase (EC 2.3.1.12)\n",
      "SSO:000002109 (0.01)\tDNA repair protein RecN\n",
      "SSO:000004554 (0.00)\tLow-specificity L-threonine aldolase (EC 4.1.2.48)\n",
      "SSO:000025018 (0.00)\tSpore coat protein CotH\n",
      "SSO:000021037 (0.00)\tPAS domain-containing protein\n",
      "SSO:000000212 (0.00)\t2-phosphosulfolactate phosphatase (EC 3.1.3.71 )\n",
      "SSO:000016804 (0.00)\tFlagellar motor protein\n",
      "SSO:000044352 (0.00)\tformate C-acetyltransferase (EC 2.3.1.54)\n",
      "SSO:000020878 (0.00)\tOuter membrane lipoprotein carrier protein LolA\n",
      "SSO:000012160 (0.00)\tChemotaxis response regulator protein-glutamate methylesterase (EC 3.1.1.61)\n",
      "SSO:000021268 (0.00)\tPatatin\n",
      "SSO:000017255 (0.00)\tGlutamate synthase\n",
      "SSO:000033511 (0.00)\tthiolase family protein\n",
      "SSO:000017940 (0.00)\tHydrogenase maturation protease\n"
     ]
    }
   ],
   "source": [
    "### Find feature details (Only for RAST features).. Not FOR INTERPRO\n",
    "\n",
    "import json\n",
    "def find_rast_details(report_string):\n",
    "    with open (\"data/SSO_dictionary.json\") as sf:\n",
    "           sso_terms = json.load(sf)['term_hash']\n",
    "    sso_dict = dict()\n",
    "    for term in sso_terms:\n",
    "        data = sso_terms[term]\n",
    "        sso_dict[data['id']] = data['name']\n",
    "        \n",
    "    report_data = report_string.split(\",\")\n",
    "    for rast_score in report_data:\n",
    "        rast_score = rast_score.lstrip()\n",
    "        rast_id = rast_score.split(\" \")[0]\n",
    "        \n",
    "        print (rast_score + \"\\t\" + sso_dict[rast_id])\n",
    "        \n",
    "        \n",
    "report_string = \"SSO:000036573 (0.71), SSO:000024182 (0.10), SSO:000001301 (0.02), SSO:000012980 (0.01), SSO:000031056 (0.01), SSO:000020884 (0.01), SSO:000043940 (0.01), SSO:000002109 (0.01), SSO:000004554 (0.00), SSO:000025018 (0.00), SSO:000021037 (0.00), SSO:000000212 (0.00), SSO:000016804 (0.00), SSO:000044352 (0.00), SSO:000020878 (0.00), SSO:000012160 (0.00), SSO:000021268 (0.00), SSO:000017255 (0.00), SSO:000033511 (0.00), SSO:000017940 (0.00)\"\n",
    "find_rast_details(report_string)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9266d65f-f27c-4b2e-b725-38a40ef0cd4e",
   "metadata": {},
   "source": [
    "### Run multiple classifiers on gram stain data with Interpro\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "81feda9f-64c9-4565-8490-42175ad5b8a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Decision Tree\n",
      "Running Random Forest\n",
      "Running XGBoost\n",
      "Running CatBoost\n",
      "Running AdaBoost\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div id=\"datatable-container-ff2a993a-3091-11ee-b959-8244d1051800\">\n",
       "      <script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.7.0/jquery.min.js\"></script>\n",
       "      <script type=\"text/javascript\" src=\"https://cdn.datatables.net/1.13.5/js/jquery.dataTables.min.js\"></script>\n",
       "      <link rel=\"stylesheet\" type=\"text/css\" href=\"https://cdn.datatables.net/1.13.5/css/jquery.dataTables.min.css\"/>\n",
       "      <script type=\"text/javascript\">\n",
       "        $(document).ready( function () {\n",
       "            $('#BGCtable').DataTable();\n",
       "        });\n",
       "      </script>\n",
       "      <!-- Insert table below -->\n",
       "      <table border=\"1\" class=\"dataframe display\" id=\"classification\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1-score</th>\n",
       "      <th>Balanced Accuracy</th>\n",
       "      <th>Top Features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.940711</td>\n",
       "      <td>0.940548</td>\n",
       "      <td>0.940711</td>\n",
       "      <td>0.940613</td>\n",
       "      <td>0.930268</td>\n",
       "      <td>IPR000184 (0.78), IPR002543 (0.05), IPR043723 (0.02), IPR004550 (0.01), IPR007835 (0.01), IPR000109 (0.01), IPR015262 (0.01), IPR008307 (0.01), IPR009293 (0.01), IPR003561 (0.01), IPR008767 (0.01), IPR019539 (0.01), IPR008567 (0.00), IPR013570 (0.00), IPR003468 (0.00), IPR005064 (0.00), IPR036473 (0.00), IPR001765 (0.00), IPR005235 (0.00), IPR017502 (0.00)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.974308</td>\n",
       "      <td>0.975346</td>\n",
       "      <td>0.974308</td>\n",
       "      <td>0.974485</td>\n",
       "      <td>0.977839</td>\n",
       "      <td>IPR039518 (0.01), IPR003400 (0.01), IPR005653 (0.01), IPR039910 (0.01), IPR023052 (0.01), IPR010827 (0.01), IPR004474 (0.01), IPR015349 (0.01), IPR036942 (0.01), IPR030921 (0.01), IPR037066 (0.01), IPR005495 (0.01), IPR028053 (0.01), IPR012910 (0.01), IPR023707 (0.01), IPR027434 (0.01), IPR039426 (0.01), IPR038594 (0.01), IPR039565 (0.01), IPR004613 (0.01)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XGBoost</td>\n",
       "      <td>0.964427</td>\n",
       "      <td>0.965063</td>\n",
       "      <td>0.964427</td>\n",
       "      <td>0.964594</td>\n",
       "      <td>0.964040</td>\n",
       "      <td>IPR000184 (0.15), IPR005495 (0.12), IPR004474 (0.06), IPR002543 (0.04), IPR018478 (0.03), IPR042214 (0.02), IPR004570 (0.02), IPR023473 (0.02), IPR043723 (0.01), IPR020988 (0.01), IPR020059 (0.01), IPR035924 (0.01), IPR002043 (0.01), IPR025097 (0.01), IPR022641 (0.01), IPR046786 (0.01), IPR010969 (0.01), IPR036105 (0.01), IPR034762 (0.01), IPR003344 (0.01)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CatBoost</td>\n",
       "      <td>0.974308</td>\n",
       "      <td>0.975004</td>\n",
       "      <td>0.974308</td>\n",
       "      <td>0.974448</td>\n",
       "      <td>0.976206</td>\n",
       "      <td>IPR005495 (14.29), IPR003400 (9.55), IPR004474 (4.25), IPR039910 (3.32), IPR036346 (2.98), IPR003802 (2.60), IPR000184 (1.55), IPR015870 (1.33), IPR017501 (0.95), IPR006412 (0.95), IPR003423 (0.89), IPR006256 (0.75), IPR004800 (0.67), IPR023087 (0.65), IPR029050 (0.63), IPR013564 (0.63), IPR004570 (0.63), IPR017689 (0.60), IPR027463 (0.60), IPR018958 (0.56)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AdaBoost</td>\n",
       "      <td>0.962451</td>\n",
       "      <td>0.962707</td>\n",
       "      <td>0.962451</td>\n",
       "      <td>0.962541</td>\n",
       "      <td>0.959320</td>\n",
       "      <td>IPR046674 (0.01), IPR024726 (0.01), IPR000428 (0.01), IPR037274 (0.01), IPR037483 (0.01), IPR040074 (0.01), IPR000330 (0.01), IPR029095 (0.01), IPR011304 (0.01), IPR024728 (0.01), IPR045926 (0.01), IPR015330 (0.01), IPR025984 (0.01), IPR025357 (0.01), IPR014084 (0.01), IPR015919 (0.01), IPR045528 (0.01), IPR008207 (0.01), IPR007168 (0.01), IPR038901 (0.01)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "    </div>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run multiple classifiers on gram stain data with Interpro\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "#Load phenotype data with gram stain data (only gram stain phenotype here)\n",
    "phenotype_gram_stain_df = pd.read_csv(\"data/gram_stain_bacdive.tsv\", sep = \"\\t\", index_col=0)\n",
    "\n",
    "phenotype = \"gram_stain\"\n",
    "phenotype_df = phenotype_gram_stain_df[phenotype].dropna()\n",
    "merged_df = interpro_df.merge(phenotype_df, left_index=True, right_index=True, how='inner')\n",
    "X = merged_df.drop(columns=[phenotype])\n",
    "y = merged_df[phenotype]\n",
    "\n",
    "# Get classifier report\n",
    "report_df = get_classifier_report(X,y)\n",
    "\n",
    "report_html_content = generate_html_table (report_df)\n",
    "\n",
    "HTML(report_html_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8c2c7cb7-0bb9-4397-b315-f765db512562",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IPR000184 (0.78)\tBacterial surface antigen (D15)\n",
      "IPR002543 (0.05)\tFtsK domain\n",
      "IPR043723 (0.02)\tDomain of unknown function DUF5665\n",
      "IPR004550 (0.01)\tL-asparaginase, type II\n",
      "IPR007835 (0.01)\tMOFRL domain\n",
      "IPR000109 (0.01)\tProton-dependent oligopeptide transporter family\n",
      "IPR015262 (0.01)\ttRNA(Ile)-lysidine synthase, substrate-binding domain\n",
      "IPR008307 (0.01)\tUncharacterised conserved protein UCP018957\n",
      "IPR009293 (0.01)\tUncharacterised protein family UPF0478\n",
      "IPR003561 (0.01)\tMutator MutT\n",
      "IPR008767 (0.01)\tBacteriophage SPP1, head-tail adaptor\n",
      "IPR019539 (0.01)\tGalactokinase, N-terminal domain\n",
      "IPR008567 (0.00)\tBeta-keto acid cleavage enzymes\n",
      "IPR013570 (0.00)\tTranscription regulator YsiA, C-terminal\n",
      "IPR003468 (0.00)\tCytochrome c oxidase, monohaem subunit/FixO\n",
      "IPR005064 (0.00)\tBordetella uptake gene\n",
      "IPR036473 (0.00)\tMolybdopterin cofactor biosynthesis MoaD-related, C-terminal domain superfamily\n",
      "IPR001765 (0.00)\tCarbonic anhydrase\n",
      "IPR005235 (0.00)\tMetallophosphoesterase, YmdB-like\n",
      "IPR017502 (0.00)\tSortase B cell surface sorting signal\n"
     ]
    }
   ],
   "source": [
    "### Find feature details (Only for RAST features).. Not FOR INTERPRO\n",
    "\n",
    "import json\n",
    "def find_interpro_details(report_string):\n",
    "    with open (\"data/Interpro.list\") as ipro:\n",
    "           interpro_terms = ipro.readlines()\n",
    "            \n",
    "    interpro_dict = dict()\n",
    "    \n",
    "    for line in interpro_terms:\n",
    "        line = line.rstrip()\n",
    "        id =  line.split(\"\\t\")[0]\n",
    "        name = line.split(\"\\t\")[2]\n",
    "        interpro_dict[id] = name\n",
    "        \n",
    "        \n",
    "    report_data = report_string.split(\",\")\n",
    "    for interpro_score in report_data:\n",
    "        interpro_score = interpro_score.lstrip()\n",
    "        interpro_id = interpro_score.split(\" \")[0]\n",
    "        \n",
    "        print (interpro_score + \"\\t\" + interpro_dict[interpro_id])\n",
    "        \n",
    "       \n",
    "report_string = \"IPR000184 (0.78), IPR002543 (0.05), IPR043723 (0.02), IPR004550 (0.01), IPR007835 (0.01), IPR000109 (0.01), IPR015262 (0.01), IPR008307 (0.01), IPR009293 (0.01), IPR003561 (0.01), IPR008767 (0.01), IPR019539 (0.01), IPR008567 (0.00), IPR013570 (0.00), IPR003468 (0.00), IPR005064 (0.00), IPR036473 (0.00), IPR001765 (0.00), IPR005235 (0.00), IPR017502 (0.00)\"\n",
    "find_interpro_details(report_string)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "697dd011-3fe9-4f17-bebb-11db082d6e86",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
